compiler/src/iree/compiler/Codegen/Common/WorkgroupSpecializationPass.cpp

```diff
@@ -0,0 +1,260 @@
+    int64_t tileSize = tileSizes[i];
+    if (tileSize == 0 || tileSize == 1) {
+      constantOps.push_back(Value());
+      continue;
+    }
+
+    // clone the minSize op in the loop and place it before scf.if
+    AffineMinOp minOp = minSizeOps[i];
+    scf::ForOp distLoop = distLoops[i];
+    // clone the lower bound and put before the nested loops.
+    BlockAndValueMapping mapperForLB;
+    Operation *lb =
+        builder.clone(*distLoop.getLowerBound().getDefiningOp(), mapperForLB);
+
+    // Clone the affine min op for the dynamic size in the current loop and
+    // place it before the nested loops. The induction variable is replaced by
+    // the cloned lower-bound above.
+    BlockAndValueMapping mapperForIV;
+    Value iv = distLoop.getInductionVar();
+    mapperForIV.map(iv, lb->getResult(0));
+    Value size =
+        builder.clone(*minOp.getOperation(), mapperForIV)->getResult(0);
+
+    // Generate a compare op that checks the dynamic size is equal to the
+    // constant main tile size.
+    Value constant = builder.create<arith::ConstantIndexOp>(loc, tileSize);
+    constantOps.push_back(constant);
+    Value cmp = builder.create<arith::CmpIOp>(loc, arith::CmpIPredicate::eq,
+                                              size, constant);
+    cond = cond ? builder.create<arith::AndIOp>(loc, cond, cmp) : cmp;
+  }
+
+  // generate scf.if %cond then { scf.for ... } else { scf.for ... }
+  auto ifOp = builder.create<scf::IfOp>(loc, cond, /*withElseRegion=*/true);
+  ifOp.getElseBodyBuilder().clone(*outermostLoop.getOperation());
+
+  // Specialize the size operations (affine min) in each for loop.
+  for (int i = 0, e = distLoops.size(); i != e; ++i) {
+    AffineMinOp minOp = minSizeOps[i];
+    if (!minOp) {
+      assert(tileSizes[i] == 0 || tileSizes[i] == 1);
+      continue;
+    }
+    LLVM_DEBUG({
+      llvm::errs() << "Replacing ";
+      minOp.dump();
+      llvm::errs() << "with ";
+      constantOps[i].dump();
+    });
+    minOp.replaceAllUsesWith(constantOps[i]);
+  }
+  outermostLoop->moveBefore(&ifOp.getThenRegion().front().front());
+  return;
+}
+
+namespace {
+struct WorkgroupSpecializationPass
+    : public WorkgroupSpecializationBase<WorkgroupSpecializationPass> {
+  void getDependentDialects(DialectRegistry &registry) const override {
+    registry.insert<AffineDialect, IREE::HAL::HALDialect, linalg::LinalgDialect,
+                    scf::SCFDialect, tensor::TensorDialect>();
+  }
+
+  void runOnOperation() override {
+    if (!clEnableWorkgroupSpecialization) return;
+
+    IREE::HAL::ExecutableVariantOp variantOp = getOperation();
+    ModuleOp innerModule = variantOp.getInnerModule();
+    llvm::StringMap<IREE::HAL::ExecutableExportOp> entryPoints =
+        getAllEntryPoints(innerModule);
+
+    for (auto funcOp : innerModule.getOps<func::FuncOp>()) {
+      SmallVector<LoopTilingAndDistributionInfo> infoVec =
+          getTiledAndDistributedLoopInfo(funcOp);
```

Comment:
This is probably not a discussion we can have on the review but I wonder if we could make it evolve to remove dependency to `LoopTilingAndDistributionInfo` in order to do something independent of IREE and tiling.
Right now it sounds like this is load bearing in the design and this may be fine for now but it means this is something IREE specific. I think we should discuss about generalizing it.

Reply:
This is a good point. I created an issue for it. https://github.com/iree-org/iree/issues/10577